### ISSUE TITLE
Why are there no detection even if training data = validation data?

### ISSUE THREAD

**Issue by chijan-nh** [eriklindernoren/PyTorch-YOLOv3#86]
I train only 10 images to verify the behavior.
I run the training about 200 epochs, however, the detection result is none and the test result shows all APs are equal to 0 and mAP =0, where `data/samples` and the validation data are same as the training data.

Why does this problem happen?

I found the confidences are very small (<0.1) but I cannot clarify why the conf.s are much less than 0.1, while the default conf. threshold is 0.8 in train.py .

**Comment by jingtianyilong** [comment]
i tried training with train.py for 30 epochs. Loss is about 2. But the weight can't detect anything, even if i detect with the original darknet. Guess there's something wrong with the weight saving. Try training with the original darknet training modules. Works fine

**Comment by OlofHarrysson** [comment]
Try finetuning the coco weights and run the detect.py to see if you get any output. I think the mAP was incorrect and updated quite recently. Are you on the latest commit?

**Comment by chijan-nh** [comment]
> Are you on the latest commit?
Yes, I use the latest commit version.

> Guess there's something wrong with the weight saving. 
I also changed the model save method from the implementation of this repository to the built-in method of pytorch.

> Try finetuning the coco weights and run the detect.py to see if you get any output. 
Sorry, I mean "the same classes" is that the images have only a same class object, not the coco 80 classes.

**Comment by Dev2022** [comment]
@chijan-nh  I also have the same issue. Please let me know if you could solve the issue.
@OlofHarrysson  The detect.py is not returning any bounding box.

**Comment by OlofHarrysson** [comment]
@chijan-nh I could get detections with the saving/loading weights system that was implemented.

So what you are saying is that you only have one class, and that class is among the coco classes? What is the class? If you are getting crappy detections after training that might be your training that's not good

**Comment by OlofHarrysson** [comment]
@chijan-nh @Dev2022 I'm not sure if it makes a difference but in the detect.py you should change the input to the non_max_suppression to the number of classes you are using.

If you are using custom labels / dataset that might be the cause of the problems. Labels need to be in the right format for example

**Comment by chijan-nh** [comment]
@OlofHarrysson 
Thanks for comments and sorry replying late.

I use my own dataset which does not share the labels to COCO.
I tried to detect meaningful bb and tuned NMS thresholds, however, the detection FP and FN look large.
Besides, the sizes of bbs are very small.
I got some bbs using the pytorch built-in save method (above), and my labels txts may be correct; I set my labels according to https://github.com/eriklindernoren/PyTorch-YOLOv3/issues/76 .

**Comment by OlofHarrysson** [comment]
What does your dataset consist of? How many classes are there? Are you starting your label numbers from 0? Perhaps there is a mismatch for the mapping from the classes to the integer? If the bbs sizes are very small, there might be something to investigate there but I can't think of anything.

Did you write any new code? Like your own pytorch dataset class?

**Comment by chijan-nh** [comment]
My dataset consists images which have abnormal points.
The number of classes is one. The .names file is only written
```
abnormality
```
and labels txt files are written such a
```
0 0.7 0.4 0.05 0.06
```
(the bounding box coordinate and size are just examples).

I also fixed the number of filters in Convolutional Layer forward YOLO Layer, because of the num of classes is not 80, in yolov3.cfg.

About over fitting experiments, I found the loss function value is very small and the precision and recall is near to 1. However, the detect.py cannot detect correctly.

The weight of COCO can be used to detect for COCO dataset and the result is correct.

**Comment by DamanTTT** [comment]
@chijan-nh @Dev2022
Hi, I had the same problem and I just solved it yesterday.

The main reason is when the model saving weights, it would save a numpy array named header_info first. But the dtype doesn't be defined so that there is a problem when loading weight.

To fix this problem, we just need to add "dtype=np.int32" behind.

> In models.py

`class Darknet(nn.Module):`
`self.header_info = np.array([0, 0, 0, self.seen, 0], dtype=np.int32)`

**Comment by chijan-nh** [comment]
@DamanTTT 
This is a just friendly question.
Could you tell us your computing environment?
I currently can not try your solution depending on various reason...... .

**Comment by zwbdmm** [comment]
@DamanTTT 
this works for me, thanks a lot!

**Comment by DamanTTT** [comment]
@chijan-nh 
Hi, ,my computing environment is Ubuntu 16.04 and Python 3.5.2 with numpy 1.13.3 and pytorch 0.4.1. Wish the information can help you.

**Comment by wanzn** [comment]
> @DamanTTT 
Hi, i took your advice and changed code, and when i use the weights i trained to detect image,  the result is no longer 0, but still very poor (trainning stage  seems pretty good).
Did you get some valid test result?
